File: RemapReferences.rb
#@category Malware/DOS

# PARAMS ###########################################################################################

from_segment   = getScriptArgs[0] || "0000"
to_segment     = getScriptArgs[1] || "07c0"
start_offset   = getScriptArgs[2] || "8000"
end_offset     = getScriptArgs[3] || "81ff"

####################################################################################################

tx_id = currentProgram.start_transaction("Remap #{from_segment}:#{start_offset}-#{end_offset} references to #{to_segment}")
success = false

begin
  ref_manager = currentProgram.reference_manager
  start_addr  = toAddr("#{from_segment}:#{start_offset}")
  end_addr    = toAddr("#{from_segment}:#{end_offset}")

  addr = start_addr
  while addr <= end_addr
    refs = ref_manager.get_references_to(addr).to_a   # snapshot to avoid iterator mutation
    unless refs.empty?
      # Compute the equivalent address in the virus segment (07c0)
      target_addr = toAddr("#{to_segment}:#{"%04x" % addr.segment_offset}")
      refs.each do |ref|
        # Add new reference pointing to the correct segment
        ref_manager.add_memory_reference(
          ref.from_address,
          target_addr,
          ref.reference_type,   # preserve original ref type (read/write/data)
          SourceType::USER_DEFINED,
          ref.operand_index     # preserve which operand the ref belongs to
        )
        ref_manager.delete(ref)
      end
      puts "Remapped #{addr} -> #{target_addr} (#{refs.size} ref#{refs.size == 1 ? '' : 's'})"
    end
    addr = addr.next   # advance one byte at a time
  end

  success = true
ensure
  # Rollback on exception
  currentProgram.end_transaction(tx_id, success)
end

File: RemapReferences.java
// Remaps references from one address range to another.
// Useful when Ghidra doesn't know the actual value of a segment register and creates references in
// the wrong segment.
//
// @category 0Saverio/DosMalware

import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.symbol.*;
import java.util.ArrayList;
import java.util.List;

public class RemapReferences extends GhidraScript {

  @Override
  protected void run() throws Exception {
    String[] args = getScriptArgs();
    String fromStart = args.length > 0 ? args[0] : "0000:8000";
    String fromEnd = args.length > 1 ? args[1] : "0000:81ff";
    String toAddress = args.length > 2 ? args[2] : "07c0:8000";
    boolean commit = true;

    Address startAddr = toAddr(fromStart);
    Address endAddr = toAddr(fromEnd);
    Address targetBase = toAddr(toAddress);
    ReferenceManager refManager = currentProgram.getReferenceManager();

    int txId = currentProgram.startTransaction("Remap references");
    boolean success = false;

    try {
      Address addr = startAddr;
      while (addr.compareTo(endAddr) <= 0) {
        List<Reference> refs = new ArrayList<>();
        refManager.getReferencesTo(addr).forEachRemaining(refs::add);
        if (!refs.isEmpty()) {
          Address targetAddr = targetBase.add(addr.getOffset() - startAddr.getOffset());
          for (Reference ref : refs) {
            Address from = ref.getFromAddress();
            RefType type = ref.getReferenceType();
            int op = ref.getOperandIndex();
            refManager.delete(ref);
            refManager.addMemoryReference(from, targetAddr, type, SourceType.USER_DEFINED, op);
          }
          printf("Remapped %s -> %s (%d ref%s)\n", addr, targetAddr, refs.size(), refs.size() == 1 ? "" : "s");
        }
        addr = addr.add(1);
      }
      success = commit;
    } finally {
      currentProgram.endTransaction(txId, success);
    }
  }
}
